### ISSUE TITLE
Extend Rucio to generic checksum column

### ISSUE THREAD

**Issue by bari12** [rucio/rucio#2410]
Motivation
----------
Right now Rucio only supports adler32 and md5 checksums via two specific columns for these checksums. This creates problems if a community has different checksums (e.g. CMS historically used crc checksums) and wants to store them. On the dev meeting on [2019-03-28](https://indico.cern.ch/event/796966/) it was discussed to extend Rucio with a generic checksum column of string type. Multiple checksums can be added (comma-separated) to this column, always specifying the checksum type. E.g. `md5:1234,adler32:5678,crc:123`

We would have to specify a strategy in migrating to this new checksum column.

**Comment by mlassnig** [comment]
Alternatively, a json column? I don't like string parsing too much

**Comment by bziemons** [comment]
is #3538 related to this?

**Comment by bari12** [comment]
Not really. We do provide already `adler32` and `md5` columns in the database. #3538 is just about an issue that at some parts of the code `adler32` is falsly required.

**Comment by panta-123** [comment]
@bari12 , should I go with  `md5:1234,adler32:5678,crc32:91011` or JSON: `{'md5': '1234', 'adler32':'5678', 'crc32': '91011'}` ?

**Comment by bari12** [comment]
Sorry the late reply @panta-123 
To me, we should converge this to a JSON type column.

**Comment by martynia** [comment]
@bari12 , I modified multiple tables in `models.py` by adding an extra column to hold a JSON dictionary. How should a backward compatibility be maintained in code? I assume the code should check. if the schema is old `(adler32` and `md5)` or new (JSON) for a case when code is updated but the DB is not.

A current state of code is that all the tables are filled in with a new JSON checksum value, transfers run, but the checks (see below) are still made with the old columns (I keep both old and new columns for the time being)

As I have new checksums now there is time to verify them and use them for checks below. If this works I'll remove` md5` and `Adler32` column from `models.py`.

 I noticed that Rucio had a notion of `GLOBALLY_SUPPORTED_CHECKSUMS` (there are only 2 there, `Adler32` and `md5`) and `Adler32` is the preferred one. However code has` Adler32` hardwired in places, like in `upload_client` here:

[/lib/rucio/client/uploadclient.py#L735
](https://github.com/rucio/rucio/blob/9ccb37a8da4e63d1af2f86902765a2c795ec028d/lib/rucio/client/uploadclient.py#L735)

So it looks like an idea of prefered  checksum is not used here.

On the other hand `transfertool/fts3.py` requires both:

[/lib/rucio/transfertool/fts3.py
](https://github.com/rucio/rucio/blob/9ccb37a8da4e63d1af2f86902765a2c795ec028d/lib/rucio/transfertool/fts3.py#L661)


The DB migration is yet another issue, what is the procedure here ? You must have faced this type of migration in the past.

**Comment by bari12** [comment]
We have reviewed this issue as part of our regular issue review.

This issue has been open for longer than the twelve-month delivery horizon, but we are making an exception and keeping it open.

The issue is currently being worked on by @martynia 

In line with our [issue management guideline](https://rucio.cern/documentation/developer/issue_management), such exceptions are granted case by case, and no issue remains open for longer than twenty-four months. By then this issue will be resolved, re-justified, or closed and labelled `wishlist`.
